File: shortestJobFirst.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <unordered_map>
#include <map>
#include <queue>
#include<set>
using namespace std;
struct Job {
	int id;
	int arrival_time;
	int burst_time;
	Job(int id, int arrival_time, int burst_time) {
		this->id = id;
		this->arrival_time = arrival_time;
		this->burst_time = burst_time;
	}
	
};
bool comp(Job p1, Job p2) {
	if (p1.arrival_time == p2.arrival_time) {
		if (p1.burst_time == p2.burst_time) {
			return p1.id < p2.id;
		}
		else {
			return p1.burst_time < p2.burst_time;
		}
	}
	else {
		return p1.arrival_time < p2.arrival_time;
	}
}
auto com = [](Job p1, Job p2) {
	return p1.burst_time > p2.burst_time;
};
vector<int> shortestJobFirst(vector<int> id, vector<int> arrival_times, vector<int> burst_times) {
	vector<Job> schedules;
	for (int i = 0; i < arrival_times.size(); i++) {
		schedules.push_back(Job(id[i], arrival_times[i], burst_times[i]));
	}
	sort(schedules.begin(), schedules.end(), comp);
	vector<int> result;
	priority_queue <Job, vector<Job>, decltype(com)> ReadyQueue(com);
	ReadyQueue.push(schedules[0]);
	int n = id.size();
	int i = 1;
	int currentTime = 0;
	while (!ReadyQueue.empty()) {
		Job top = ReadyQueue.top();
		ReadyQueue.pop();
		result.push_back(top.id);
		currentTime += top.burst_time;
		while (i < n and schedules[i].arrival_time <= currentTime) {
			ReadyQueue.push(schedules[i]);
			i++;
		}
	}
	
	return result;
}

int main() {
	vector<int> id = { 1, 2, 3, 4 };
	vector<int> arri = { 2, 0, 4, 5 };
	vector<int> burst = { 3, 4, 2, 4 };
	vector<int> result = shortestJobFirst(id, arri, burst);
	for (auto i : result) {
		cout << i << " ";
	}
	
    
	
	return 0;
}

File: shortestJobFirst.py
from collections import namedtuple
from operator import attrgetter
import heapq


def shortest_job_first(ids, arrival_times, burst_times):
    n = len(id)
    jobs = []
    job_attributes = ['arrival_time', 'burst_time', 'id']
    Job = namedtuple('Job', job_attributes)
    for job_details in zip(arrival_times, burst_times, ids):
        jobs.append(Job(*job_details))
    # We then first sort by arrival_time, on tie we
    # sort by burst_time, on tie we sort by id
    jobs = sorted(jobs, key=attrgetter('arrival_time', 'burst_time', 'id'))
    ready_queue = []
    heapq.heappush(ready_queue, [jobs[0].burst_time, jobs[0].id, jobs[0].arrival_time])
    i = 1  # id number
    current_time = 0
    result = []
    while ready_queue:
        top_element = heapq.heappop(ready_queue)
        result.append(top_element[1])  # appending id
        current_time += top_element[0]  # appending burst time
        while i < len(arrival_times) and jobs[i].arrival_time <= current_time:
            # python heap sort by first element
            # that's why we are pushing burst time first
            heapq.heappush(ready_queue, [jobs[i].burst_time, jobs[i].id, jobs[i].arrival_time])
            i += 1
    return result


id = [1, 2, 3, 4]
arrival = [2, 0, 4, 5]
burst = [3, 4, 2, 4]
print(shortest_job_first(id, arrival, burst))
